### ISSUE TITLE
Mocking abstract classes

### ISSUE THREAD

**Issue by ssynix** [NagRock/ts-mockito#25]
I see that this is not supported yet:
> 'Argument of type 'typeof AbstractClass' is not assignable to parameter of type 'new (...args: any[]) => AbstractClass'.
  Cannot assign an abstract constructor type to a non-abstract constructor type.'

Can this be supported in TypeScript or not yet possible?

Thanks!

**Comment by michalstocki** [comment]
Mocking abstract classes is not a trivial task, because abstract methods do not exist in JavaScript. We can't simply iterate over the object properties to create the correct mock object.
```typescript
abstract class AbstractClass {
  public abstract getValue():number;

  public getOtherValue():string {
    return '';
  }
}
```
produces:
```javascript
    var AbstractClass = (function () {
        function AbstractClass() {
        }
        AbstractClass.prototype.getOtherValue = function () {
            return '';
        };
        return AbstractClass;
    }());
```
This could potentially be implemented in the way similar to mocking of interfaces (using Proxy object). Maybe using the same technique. However the feature is still not supported. @NagRock could you confirm whether the feature is planned in the nearest future?
```
mock:AbstractClass = mockType<AbstractClass>();
```

**Comment by michalstocki** [comment]
Nevertheless, the simple workaround for lack of this feature is using any extension of the `AbstractClass` to mock;
```typescript
// Exntension.ts
export class Extension extends AbstractClass {
  public getValue():number {
    return 1;
  }
}

// test
const abstractMock:AbstractClass = mock(Extension);
```
Note, that `abstractMock` is in type of `AbstractClass` to not allow calling of extra methods that `Extenstion` could have.

**Comment by ssynix** [comment]
Thanks for the explanation and the quick response! I'm using the workaround right now. One thing I don't like about it is that any changes in the number or signature of abstract methods has to be replicated in the Extension class.

**Comment by michalstocki** [comment]
😄 But it's the essential value of an abstract class: it enforces consistency of it's implementations. I think you should not create a special Extension class only for the mocking purpose. You can use one of the extensions already existing in your code.

**Comment by NagRock** [comment]
@michalstocki mocking interfaces is planned but rather not in the coming weeks. Also I will need some help with this feature from you.

**Comment by NagRock** [comment]
Thanks for @dyong0 we can now mock `abstract` and `generic` types!
I will close this issue and open new for interfaces mocking.

**Comment by NagRock** [comment]
@ssynix please give feedback is it works as expected.

**Comment by dyong0** [comment]
@NagRock I found another issue on mocking generics. Mocking a generic class or interface loses members with its generic types. For example, mongoose.Model is an generic interface
```js
interface Model<T extends Document> extends NodeJS.EventEmitter, ModelProperties { ... }
```
which has
```js
create(docs: any[], callback?: (err: any, res: T[]) => void): Promise<T[]>;
where(path: string, val?: Object): Query<any>;
```
When you mock it:
```js
let mockedModel = mock(mongoose.Model);
let model = instance(mockedModel);
```
it loses some parts of its body:
![image](https://user-images.githubusercontent.com/7734590/29369622-117ac9c4-82de-11e7-9323-582dd2e73385.png)
yet keeps members without its generic types:
![image](https://user-images.githubusercontent.com/7734590/29369684-45f67978-82de-11e7-8092-7c24bace3c80.png)

Can you please add a note for this issue on mocking generics? README would be nice to place it.

**Comment by dyong0** [comment]
Plus, when you `console.log()` mocked a generic. It displays `null`, but actually it's not.
```js
let mockedModel = mock(mongoose.Model);
let model = instance(mockedModel);

console.log(model); // null
model === null; // false
```

**Comment by NagRock** [comment]
@dyong0 I will try to fix it instead of adding info in readme. Thanks for reporting this!

**Comment by NagRock** [comment]
@dyong0 you trying to mock interface (not abstract class) that is not supported yet.

**Comment by dyong0** [comment]
@NagRock right, it doesn't have tests for mocking interfaces. but looks working fine for interfaces too. btw the issue i reported was about mocking generics. i think the same problem happens on generic classes as well as abstract ones.

**Comment by NagRock** [comment]
@dyong0 could you please provide a sample repository with this? It would be much simpler for me to fix this with project that reproduces issue.

**Comment by mpiroc** [comment]
When mocking abstract classes, abstract methods are not mocked (version 2.1.1). I assume that this is the same problem that prevents mocking interfaces, but at minimum the docs should be updated to prevent confusion.

```typescript
abstract class Foo {
  abstract foo(): void;
  bar(): void {}
}

...

const mock = mock(Foo)
mock.foo // undefined
mock.bar // function
```

**Comment by assafaloni** [comment]
I tried to run @mpiroc example with v2.2.5 and got the same output (foo is undefined),
I did include es6 for tsconfig lib param.

I dont understand - this should be supported now? if so what am i doing wrong?
Thanks!

**Comment by NagRock** [comment]
Hi, thanks for reporting. I will check it.

**Comment by julianosam** [comment]
Hey guys. Was this issue solved? I can see the PR that adds support for mocking abstract classes was merged, but I'm still getting the error mentioned by @mpiroc. I'm using version 2.2.9. Thanks

**Comment by woodcockjosh** [comment]
When we gonna be able to mock interfaces? It's like really I have to mock an implementation? That's totally broken and defeats the whole point of mocking.

**Comment by pbadenski** [comment]
@woodcockjosh it doesn't seem to be that obvious how could this be implemented easily.. I use this workaround:

```typescript
interface LoadBalancer {
  calculate: (...) => Promise<any>;
}

const loadBalancerMock = mock.mock<LoadBalancer>({ prototype: { calculate: (): any => undefined } } as any);```

**Comment by woodcockjosh** [comment]
Fine, I'll do a pull request with a solution. That's obviously not the solution.

**Comment by woodcockjosh** [comment]
Sorry, you know developers write before they think sometimes :-/. I meant I
think there's a better solution and will try to contribute with more than a
complaint.

**Comment by joaovieira** [comment]
+1 for mocking interfaces! 😩
